### DISCUSSION TITLE
Method not implemented: PageHandler.endListBand while generating an Excel report

### DISCUSSION THREAD

**Discussion by rt15** [eclipse-birt/birt#1427] [Q&A] [answered]
Hello,
The product I am working on is using Birt to generate some reports.
But we are facing an exception when we generate an Excel (xlsx) report with one of our template.
Other output formats are okay with this template though.

I am not familiar with Birt, but here are my investigations.
All the classes that I will be talking about are in Birt engine, in uk.co.spudsoft.birt.emitters.excel folder.
The error message is triggered by AbstractHandler because, indeed, PageHandler does not implement endListBand.
And I suppose that this is expected. There should not be a list band directly inside a page.
A list band should be in a list.
And actually, when the error occurs, we are in a list (we are after a call to startList, and before a call to endList).
So the current handler should be TopLevelListHandler.
But it is not.
Why?
Because there was a list in the list.

Here is what is going on:
startList is called while the handler is PageHandler.
So PageHandler create a new TopLevelListHandler and set it as current handler.
startList is called again, because of a nested list. This time, TopLevelListHandler deals with it. It does not create a new handler.
Then endList is called as we reach the end of the nested list.
It is handled by TopLevelListHandler (in fact its parent, AbstractRealListHandler).
AbstractRealListHandler restores the PageHandler as current handler (state.setHandler(parent)).
But that feels like a mistake, since we are still in a list.
So later, when endListBand is called, the current handler is PageHandler... And the error occurs.

The report template looks like this:
```xml
<body>
	<list id="970">
		<property name="dataSet">Master</property>
		<list-property name="boundDataColumns">
			<structure>
				...
			</structure>
		</list-property>
		<header>
			<label id="347">
				...
			</label>
		</header>
		<detail>
			<data id="971">
				...
			</data>
			<list id="972">
				<detail>
					<table id="996">
						...
```
We can see a list in a list.

I don't know if there is a limitation in Birt and that we should not be able to nest lists or if this is a bug in Excel emitter.
If it is a bug in Excel emitter, I don't know if TopLevelListHandler should create a new current handler when startList is called, or if TopLevelListHandler should restore the parent in endList only if we are not in a nested list (to do so, TopLevelListHandler should count the number of calls to startList).

[edit]
I built a relatively simple report template (and two xml data sources) that allows to reproduce the issue easily.
[The report template](https://github.com/rt15/temp_birt_template/blob/main/NestedList.rpttemplatel).
[First xml file.](https://github.com/rt15/temp_birt_template/blob/main/CustomersDataSource.xml)
[Second xml file.](https://github.com/rt15/temp_birt_template/blob/main/OperationsDataSource.xmll)

Functional description:
The first xml file contains a list of customers.
The second xml file contains a list of operations (Customer, Account, Amount).
In the report template, a first list loop through the customers.
And a second nested list displays the operations of the current customer.

I reproduced with:
birt-rcp-report-designer-4.13.0-20230302-win32.win32.x86_64.zip

Interestingly, I also tried with:
birt-rcp-report-designer-4.8.0
In addition to "View Report as XLS_SPUDSOFT" and "View Report as XLSX" that does not work just like in 4.13.0, there is an additional possibility, "View Report as XLS", which works just fine.

Here is a screenshot of the resulting report:
![Report with old XLS emitter](https://raw.githubusercontent.com/rt15/temp_birt_template/main/Expected.jpg)

It would be nice if xls spudsoft would be able to produce xls and xlsx file just like this one.

**Comment by speckyspooky** [comment]
I didn't understand you case exactly for that we would need a simple example to verify it.
But I can confirm list-elements are special because it is really dependent of your list implementation like html or something else.
In some cases lists will be handled like tables and the excel emitter try to parse the list-tables to excel tables.
But you write about nested lists and the excel-emitter cannot transform nested tables or nested lists.
Because the count of the largest defined table define the layout at the start point of creation
and if the emitter find a nested table/list he cannot add the new structure to the internal layout correctly because the layout was already defined before.

But for a detailed analysis or may be emitter enhancement we would need like mentioned before a simple example without external data sources.

**Reply by rt15** [reply]
I added everything to reproduce fairly easily in my original question.

**Reply by speckyspooky** [reply]
@rt15 Thanks for your example. It is really good and I can reproduce the issue.
The topic is that the list-object isn't full implemented because the development was concentrated to handle tables & grids.
I will take a deeper look into it.

**Comment by hvbtup** [comment]
You should also try the latest release from https://github.com/triestram-partner/birt - this is a fork where I used changes from the original author of the "Spudsoft emitter". Maybe it works with that fork. If it does, this might help in identifying the cause.

**Reply by speckyspooky** [reply]
The report will cause the same issue at your fork. I checkt the involved classes.
But with the merge of the new PR to your fork it will work too.

**Comment by speckyspooky** [comment, marked as answer]
I have created th PR #1433 which includes the improvement of spudsoft xls/xlsx to show the list element with the table
based on the given test report.

### Screen with the result of the PR:
![2023-09-23 16_01_05-NestedList(1) xlsx  -  Geschützte Ansicht - Excel](https://github.com/eclipse-birt/birt/assets/41593722/c3606700-2af9-40b8-a8ac-379bcd3decff)

### Test report with small formating changes to test the excel layout
[NestedList.zip](https://github.com/eclipse-birt/birt/files/12706616/NestedList.zip)

### XLSX & XLS - result
[NestedList.xlsx](https://github.com/eclipse-birt/birt/files/12706617/NestedList.xlsx)
[NestedList.xls](https://github.com/eclipse-birt/birt/files/12706618/NestedList.xls)

**Reply by rt15** [reply]
Thank you very much for your reactivity and the pull request!
